pulumi/python/automation/main.py

```diff
@@ -170,19 +209,34 @@ def main():
         provider.display_execution_order(output=sys.stdout)
         sys.exit(0)
 
-    # For the other operations, we need the configuration files parsed, so we do the parsing upfront.
-
+    # We parse the environment file up front in order to have the necessary values required by this program.
+    # The logic around the PULUMI_STACK accounts for three scenarios:
+    #
+    # 1. If there is no environment file, the argument given on the CLI is used and added to the environment file.
+    # 2. If there is a difference between the CLI and the environment file, the environment file value is used.
+    # 3. If there is an environment file with no PULUMI_STACK, the environment file is appended with the argument.
     try:
         env_config = env_config_parser.read()
     except FileNotFoundError as e:
-        msg = 'Environment configuration file not found. This file must exist at the path: %s'
-        RUNNER_LOG.error(msg, e.filename)
+        # No file, we create one and then read it back in
+        write_env(e, stack_name)
+        env_config = env_config_parser.read()
+
+    if env_config.stack_name() is None:
+        # Found file, if there is no stack we append it
+        try:
+            env_config = env_config_parser.read()
+        except FileNotFoundError:
+            sys.exit(2)
+        append_env(env_config, stack_name)
```

Comment:
Question: if we found a file of the same name but id doesn't contain a stack identifier, would it be simpler to to a `write_env` instead?  That way we always know that the file we wind up with contains the `env_config` we expect?  With append I can see some danger of creating a file that has unexpected stuff.  

If there's worry about blowing away some existing config, the file could be backed up before this occurs, or we could prompt the user to let us delete it, or quit the process and deal with it themselves.

I'm lacking a lot of context, but in general my point is that appending to an unknown existing file could create unpredictable edge cases, whereas forcing an explicit write of the incoming config to the file would give us more assurance moving forward that we are working with something known.

Reply:
This is a good point; the case I'm seeing this useful for is one where we decide "we need to pass environment variables into the process through the environment file". 

In this case, PULUMI_STACK is just one of the potential variables that would be applied, hence we append. My thought is that if the text gets appended to a file that has garbage in it, the parser will either just ignore it or blow up which *should* cause the parser to fail.

However, let me test both of the above cases to see if I can get a real answer to those questions.